libc/unistd_file.c

```diff
@@ -0,0 +1,774 @@
+
+	/* repoen file readonly */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+	fd = open(filename, O_RDONLY);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fd);
+	{
+		/* try to read LINE1 from file */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, read(fd, buf, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_STRING(LINE1, buf);
+
+		/* assert EOF */
+		TEST_ASSERT_EQUAL_INT(0, read(fd, buf, sizeof(LINE1) - 1));
+		TEST_ASSERT_EMPTY(buf);
+	}
+}
+
+
+TEST(unistd_file, file_write_zero)
+{
+	/* test writing no bytes */
+	TEST_ASSERT_EQUAL_INT(0, write(fd, NULL, 0));
+}
+
+/* simultaneous write with two open descriptors */
+TEST(unistd_file, file_write_reopened)
+{
+	int sum;
+
+	/* reopen file descriptor and open another for reading */
+	fd2 = open(filename, O_WRONLY | O_CREAT | O_APPEND);
+	int fdr = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from two different file descriptors */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd2, LINE2, sizeof(LINE2) - 1));
+
+		/* try to read too much for simple error detection */
+		sum = (sizeof(LINE1) - 1) + (sizeof(LINE2) - 1);
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, read(fdr, buf, sum));
+		TEST_ASSERT_EQUAL_STRING(LINE2, buf);
+
+		/* test for EOF */
+		TEST_ASSERT_EQUAL_INT(0, read(fdr, buf, sum));
+		TEST_ASSERT_EMPTY(buf);
+	}
+	TEST_ASSERT_EQUAL_INT(0, close(fdr));
+	TEST_ASSERT_EQUAL_INT(0, close(fd2));
+}
+
+
+/* test simultaneous write() on duplicated file descriptor */
+TEST(unistd_file, file_write_dup)
+{
+	int sum;
+
+	/* duplicate file descriptor and open another for reading */
+	fd2 = dup(fd);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd2);
+	int fdr = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from descriptor and its copy */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd2, LINE2, sizeof(LINE2) - 1));
+
+		/* try to read too much for simple error detection */
+		sum = (sizeof(LINE1) - 1) + (sizeof(LINE2) - 1);
+		TEST_ASSERT_EQUAL_INT(sum, read(fdr, buf, sum));
+		TEST_ASSERT_EQUAL_STRING_LEN(LINE1, &buf[0], sizeof(LINE1) - 1);
+		TEST_ASSERT_EQUAL_STRING(LINE2, &buf[sizeof(LINE1) - 1]);
+	}
+}
+
+/* Test write() to closed descriptor */
+TEST(unistd_file, file_readwrite_badfd)
+{
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	TEST_ASSERT_EQUAL_INT(-1, write(fd, LINE1, sizeof(LINE1) - 1));
+	TEST_ASSERT_EQUAL_INT(-1, read(fd, buf, sizeof(LINE1) - 1));
+
+	/* this test closes fd by itself, invalidate it */
+	fd = -1;
+}
+
+/* test write() outside of the file */
+TEST(unistd_file, file_write_efbig)
+{
+	/* set offset in file at some positive value */
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, lseek(fd, 1, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(-1, write(fd, LINE2, sizeof(LINE2)));
```

Comment:
Hmm, why does this fail? Shouldn't this write just increase file size?

Reply:
I think it only talks about offsets being too large (larger than limits). Would this assert be true on glibc?

Reply:
It seems that the case where we start outside a file is included, because it failed on host. I've corrected this case, but have these changes locally yet.